gradual type checking with mypy

It'd be nice to have some type checking. I just made a [silly mistake](https://github.com/goavki/apertium-apy/commit/988fb1ae9fc2fb1a20acb4e00cf1053c27fee3d6) when renaming a variable (sending the *module* named html into re.sub instead of the html-contents variable …), which a good type checker would've caught right away. Python can't do very good type checking, but at least *some* type checking is possible with mypy, and `mypy --check-untyped-defs` would've caught this.

We'd have to do any type annotations in comments instead of the nice inline [PEP484](https://www.python.org/dev/peps/pep-0484/) syntax, since we want to support older Python versions.

We might have to only enable it for newer python versions in travis.

We might have to include some stubs for modules (`servlet.py:55: error: Cannot find module named 'cld2full'`), or tell mypy to ignore some things (`translation.py:10: error: Name 'locks' already defined` because we fallback to toro.locks in a try-except) or fill in for some bugs in mypy (`translation.py:12: error: Module 'select' has no attribute 'PIPE_BUF'`).

Remaining warnings after 89c44a0:
```
$ mypy --check-untyped-defs servlet.py              
servlet.py:23: error: No library stub file for module 'tornado.locks'
servlet.py:23: note: (Stub files are from https://github.com/python/typeshed)
servlet.py:25: error: No library stub file for module 'tornado'
servlet.py:26: error: No library stub file for module 'tornado.web'
servlet.py:27: error: No library stub file for module 'tornado.httpserver'
servlet.py:28: error: No library stub file for module 'tornado.httputil'
servlet.py:29: error: No library stub file for module 'tornado.process'
servlet.py:30: error: No library stub file for module 'tornado.iostream'
servlet.py:34: error: No library stub file for module 'tornado.escape'
servlet.py:36: error: No library stub file for module 'tornado.log'
servlet.py:38: error: No library stub file for module 'tornado.options'
servlet.py:53: error: Cannot find module named 'cld2full'
servlet.py:53: note: (Perhaps setting MYPYPATH or using the "--ignore-missing-imports" flag would help)
servlet.py:61: error: No library stub file for module 'chardet'
toro.py:26: error: No library stub file for module 'tornado.concurrent'
keys.py:8: error: Cannot infer type argument 2 of "defaultdict"
toro.py:177: error: Need type annotation for variable   
toro.py:245: error: Need type annotation for variable   
toro.py:382: error: Need type annotation for variable                                                             
toro.py:384: error: Need type annotation for variable                                                             
toro.py:388: error: Need type annotation for variable       
translation.py:10: error: Name 'locks' already defined                                                                                  
translation.py:12: error: Module 'select' has no attribute 'PIPE_BUF'
translation.py:103: error: Need type annotation for variable
translation_py32.py:10: error: Name 'locks' already defined
translation_py32.py:12: error: Module 'select' has no attribute 'PIPE_BUF'
translation_py32.py:103: error: Need type annotation for variable
missingdb.py:16: error: Need type annotation for variable                  
missingdb.py:18: error: Need type annotation for variable
systemd.py:43: error: Need type annotation for variable
util.py:22: error: Need type annotation for variable
modeSearch.py:35: error: Need type annotation for variable
servlet.py:17: error: Module 'multiprocessing' has no attribute 'TimeoutError'
servlet.py:50: error: Name 'translation' already defined
servlet.py:87: error: None has no attribute "commit"             
servlet.py:90: error: None has no attribute "commit"                                                                                    
servlet.py:91: error: None has no attribute "closeDb"
servlet.py:97: error: Need type annotation for variable
servlet.py:98: error: Need type annotation for variable
servlet.py:99: error: Need type annotation for variable                                                                                 
servlet.py:100: error: Need type annotation for variable
servlet.py:101: error: Need type annotation for variable
servlet.py:102: error: Need type annotation for variable
servlet.py:103: error: Need type annotation for variable
servlet.py:104: error: Need type annotation for variable
servlet.py:110: error: Need type annotation for variable
servlet.py:114: error: Need type annotation for variable
servlet.py:123: error: Need type annotation for variable
servlet.py:272: error: All conditional function variants must have identical signatures
servlet.py:362: error: None has no attribute "noteUnknown"
servlet.py:976: error: Need type annotation for variable
servlet.py:977: error: Need type annotation for variable
servlet.py:978: error: Need type annotation for variable
servlet.py:979: error: Need type annotation for variable
servlet.py:1203: error: Incompatible types in assignment (expression has type IO[Any], variable has type "TextIO")
servlet.py:1204: error: Incompatible types in assignment (expression has type IO[Any], variable has type "TextIO")
servlet.py:1213: error: "module" has no attribute "handlers"

```

--- a/.gitignore
+++ b/.gitignore
@@ -3,3 +3,4 @@
 /__pycache__/
 /langNames.db
 /cache/
+/.mypy_cache/

--- a/.travis.yml
+++ b/.travis.yml
@@ -13,9 +13,13 @@ python:
     - "nightly"
 
 install:
-    - if [[ $TRAVIS_PYTHON_VERSION = 3.2 ]]; then pip3 install tornado==4.3; fi
-    - if [[ $TRAVIS_PYTHON_VERSION != 3.2 ]]; then pip3 install tornado; fi
-    - pip3 install flake8 mypy
+    - if [[ $TRAVIS_PYTHON_VERSION = 3.2 ]]; then
+        pip3 install tornado==4.3;
+      fi
+    - if [[ $TRAVIS_PYTHON_VERSION != 3.2 ]]; then
+        pip3 install tornado mypy;
+      fi
+    - pip3 install flake8
 
 before_script:
     - wget http://apertium.projectjj.com/apt/install-nightly.sh -O - | sudo bash
@@ -25,8 +29,17 @@ before_script:
     - ( cd /tmp/languages/apertium-nno && ./autogen.sh && make -j2 )
     # flake8 fails to run at all under 3.2 on Travis
     # - if [[ $TRAVIS_PYTHON_VERSION = 3.2 ]]; then flake8 . --max-line-length 180 --exclude=toro.py,translation.py; fi
-    - if [[ $TRAVIS_PYTHON_VERSION != 3.2 ]]; then flake8 . --max-line-length 180 --exclude=toro.py,translation_py32.py; fi
-    - lines=$(mypy --check-untyped-defs --strict-optional **/*.py | grep -v 'note:\|No library stub file'); exit $(echo "$lines" | wc -l)
+    - if [[ $TRAVIS_PYTHON_VERSION != 3.2 ]]; then
+        flake8 . --max-line-length 180 --exclude=toro.py,translation_py32.py;
+      fi
+    - if [[ $TRAVIS_PYTHON_VERSION != 3.2 ]]; then
+        lines="$(mypy --check-untyped-defs --strict-optional ./**/*.py)";
+        echo "$lines";
+        lines="$(echo "$lines" | grep -v 'note:\|No library stub file')";
+        count="$(echo "$lines" | wc -l)"
+        echo "mypy reported $count critical errors"
+        exit "$count";
+      fi
 
 script:
     - NONPAIRS=/tmp/languages t/run-tests

--- a/.travis.yml
+++ b/.travis.yml
@@ -15,7 +15,7 @@ python:
 install:
     - if [[ $TRAVIS_PYTHON_VERSION = 3.2 ]]; then pip3 install tornado==4.3; fi
     - if [[ $TRAVIS_PYTHON_VERSION != 3.2 ]]; then pip3 install tornado; fi
-    - pip3 install flake8
+    - pip3 install flake8 mypy
 
 before_script:
     - wget http://apertium.projectjj.com/apt/install-nightly.sh -O - | sudo bash
@@ -26,6 +26,7 @@ before_script:
     # flake8 fails to run at all under 3.2 on Travis
     # - if [[ $TRAVIS_PYTHON_VERSION = 3.2 ]]; then flake8 . --max-line-length 180 --exclude=toro.py,translation.py; fi
     - if [[ $TRAVIS_PYTHON_VERSION != 3.2 ]]; then flake8 . --max-line-length 180 --exclude=toro.py,translation_py32.py; fi
+    - lines=$(mypy --check-untyped-defs --strict-optional **/*.py | grep -v 'note:\|No library stub file'); exit $(echo "$lines" | wc -l)
 
 script:
     - NONPAIRS=/tmp/languages t/run-tests

--- a/gateway.py
+++ b/gateway.py
@@ -10,7 +10,7 @@ import random
 import socket
 import servlet
 import pprint
-from typing import Any, Dict, List, Set, Tuple  # noqa
+from typing import Any, Dict, List, Set, Tuple  # noqa: F401
 from collections import OrderedDict
 import tornado
 import tornado.httpserver

--- a/gateway.py
+++ b/gateway.py
@@ -1,5 +1,10 @@
 #!/usr/bin/env python3
 
+try:
+    from typing import Any, Dict, List, Set, Tuple  # noqa: F401
+except ImportError:  # 3.2
+    pass
+
 import argparse
 import logging
 import sys
@@ -10,7 +15,6 @@ import random
 import socket
 import servlet
 import pprint
-from typing import Any, Dict, List, Set, Tuple  # noqa: F401
 from collections import OrderedDict
 import tornado
 import tornado.httpserver

--- a/gateway.py
+++ b/gateway.py
@@ -10,6 +10,7 @@ import random
 import socket
 import servlet
 import pprint
+from typing import Any, Dict, List, Set, Tuple
 from collections import OrderedDict, defaultdict
 import tornado
 import tornado.httpserver
@@ -19,7 +20,7 @@ from tornado.web import RequestHandler
 try:  # 3.1
     from tornado.log import enable_pretty_logging
 except ImportError:  # 2.1
-    from tornado.options import enable_pretty_logging
+    from tornado.options import enable_pretty_logging # type: ignore
 
 global verifySSLCert
 
@@ -43,13 +44,17 @@ class requestHandler(RequestHandler):
     def get(self):
         path = self.request.path
         mode, langPair, perWordModes = [None] * 3
-        pathToMode = defaultdict(lambda: None,
-                                 {'/translate': 'pairs', '/analyze': 'analyzers',
-                                  '/analyse': 'analyzers', '/generate': 'generators',
-                                  '/perWord': 'perWord', '/coverage': 'coverage',
-                                  '/listLanguageNames': 'languageNames', '/identifyLang': 'identifyLang',
-                                  '/getLocale': 'getLocale'
-                                  })
+        pathToMode = {
+            '/translate': 'pairs', '/analyze': 'analyzers',
+            '/analyse': 'analyzers', '/generate': 'generators',
+            '/perWord': 'perWord', '/coverage': 'coverage',
+            '/listLanguageNames': 'languageNames', '/identifyLang': 'identifyLang',
+            '/getLocale': 'getLocale'
+        }
+
+        if not path in pathToMode:
+            return self.send_error(400)
+
         mode = pathToMode[path]
 
         if path == '/translate':
@@ -64,9 +69,6 @@ class requestHandler(RequestHandler):
             perWordModes = self.get_argument('modes').split()
         elif path == '/coverage':
             langPair = self.get_argument('mode')
-
-        if not mode:
-            return self.send_error(400)
 
         query = self.request.query
         headers = self.request.headers
@@ -276,23 +278,23 @@ class Fastest(Balancer):
             modeToURL = {'pairs': 'translate', 'generators': 'generate', 'analyzers': 'analyze', 'taggers': 'tag', 'coverage': 'analyze'}
             if mode in modeToURL:
                 if (modeToURL[mode], langPair) in self.serverlist:
-                    possibleServers = list(self.serverlist[(modeToURL[mode], langPair)])
-                    if len(possibleServers):
-                        return possibleServers[0]
+                    possibleServersList = list(self.serverlist[(modeToURL[mode], langPair)])
+                    if len(possibleServersList):
+                        return possibleServersList[0]
             elif mode == 'languageNames' or mode == 'identifyLang' or mode == 'getLocale':
                 return next(self.serverCycle)
             elif mode == 'perWord':
                 modes = kwargs['perWordModes']
-                possibleServers = set()
+                possibleServersSet = set() # type: Set
                 if ('morph' in modes or 'biltrans' in modes) and ('analyze', langPair) in self.serverlist:
-                    possibleServers.update(self.serverlist[('analyze', langPair)])
+                    possibleServersSet.update(self.serverlist[('analyze', langPair)])
                 elif ('tagger' in modes or 'disambig' in modes or 'translate' in modes) and ('tag', langPair) in self.serverlist:
-                    if possibleServers:
-                        possibleServers &= self.serverlist[('tag', langPair)]
+                    if possibleServersSet:
+                        possibleServersSet &= self.serverlist[('tag', langPair)]
                     else:
-                        possibleServers.update(self.serverlist[('tag', langPair)])
-                if len(possibleServers):
-                    return list(possibleServers)[0]
+                        possibleServersSet.update(self.serverlist[('tag', langPair)])
+                if len(possibleServersSet):
+                    return list(possibleServersSet)[0]
         else:
             logging.critical('Empty serverlist')
             sys.exit(-1)
@@ -373,7 +375,7 @@ def testServerPool(serverList):
         '/list?q=taggers': lambda x: isinstance(x, dict) and all(map(lambda y: isinstance(y, str), list(x.keys()) + list(x.values()))),
         '/list?q=generators': lambda x: isinstance(x, dict) and all(map(lambda y: isinstance(y, str), list(x.keys()) + list(x.values())))
     }
-    testResults = {server: {} for server in serverList}
+    testResults = {server: {} for server in serverList} # type: Dict[Any, Dict[str, Tuple[bool, float]]]
     http = tornado.httpclient.HTTPClient()
 
     def handleResult(result, test, server):
@@ -424,7 +426,7 @@ def determineServerCapabilities(serverlist):
     # look at the return codes in order to do this.
     http = tornado.httpclient.HTTPClient()
     modes = ("pairs", "taggers", "generators", "analyzers")
-    capabilities = {}
+    capabilities = {} # type: Dict[str, Dict]
     for (domain, port) in serverlist:
         server = (domain, port)
         for mode in modes:

--- a/gateway.py
+++ b/gateway.py
@@ -10,8 +10,8 @@ import random
 import socket
 import servlet
 import pprint
-from typing import Any, Dict, List, Set, Tuple
-from collections import OrderedDict, defaultdict
+from typing import Any, Dict, List, Set, Tuple  # noqa
+from collections import OrderedDict
 import tornado
 import tornado.httpserver
 import tornado.web
@@ -20,7 +20,7 @@ from tornado.web import RequestHandler
 try:  # 3.1
     from tornado.log import enable_pretty_logging
 except ImportError:  # 2.1
-    from tornado.options import enable_pretty_logging # type: ignore
+    from tornado.options import enable_pretty_logging  # type: ignore
 
 global verifySSLCert
 
@@ -52,7 +52,7 @@ class requestHandler(RequestHandler):
             '/getLocale': 'getLocale'
         }
 
-        if not path in pathToMode:
+        if path not in pathToMode:
             return self.send_error(400)
 
         mode = pathToMode[path]
@@ -285,7 +285,7 @@ class Fastest(Balancer):
                 return next(self.serverCycle)
             elif mode == 'perWord':
                 modes = kwargs['perWordModes']
-                possibleServersSet = set() # type: Set
+                possibleServersSet = set()  # type: Set
                 if ('morph' in modes or 'biltrans' in modes) and ('analyze', langPair) in self.serverlist:
                     possibleServersSet.update(self.serverlist[('analyze', langPair)])
                 elif ('tagger' in modes or 'disambig' in modes or 'translate' in modes) and ('tag', langPair) in self.serverlist:
@@ -375,7 +375,7 @@ def testServerPool(serverList):
         '/list?q=taggers': lambda x: isinstance(x, dict) and all(map(lambda y: isinstance(y, str), list(x.keys()) + list(x.values()))),
         '/list?q=generators': lambda x: isinstance(x, dict) and all(map(lambda y: isinstance(y, str), list(x.keys()) + list(x.values())))
     }
-    testResults = {server: {} for server in serverList} # type: Dict[Any, Dict[str, Tuple[bool, float]]]
+    testResults = {server: {} for server in serverList}  # type: Dict[Any, Dict[str, Tuple[bool, float]]]
     http = tornado.httpclient.HTTPClient()
 
     def handleResult(result, test, server):
@@ -411,7 +411,7 @@ def determineServerCapabilities(serverlist):
 
     The return data from this function is a little complex, better illustrated than described:
     capabilities = {
-        "pairs": { #note that pairs is a special mode compared to taggers/generators/analyzers
+        "pairs": {  #note that pairs is a special mode compared to taggers/generators/analyzers
             ("lang", "pair"): [(server1, port1), (server2, port2)]
             }
         "taggers|generators|analyzers": {
@@ -426,7 +426,7 @@ def determineServerCapabilities(serverlist):
     # look at the return codes in order to do this.
     http = tornado.httpclient.HTTPClient()
     modes = ("pairs", "taggers", "generators", "analyzers")
-    capabilities = {} # type: Dict[str, Dict]
+    capabilities = {}  # type: Dict[str, Dict]
     for (domain, port) in serverlist:
         server = (domain, port)
         for mode in modes:

--- a/keys.py
+++ b/keys.py
@@ -1,4 +1,10 @@
-from typing import Dict  # noqa: F401
+#!/usr/bin/env python3
+
+try:
+    from typing import Dict  # noqa: F401
+except ImportError:  # 3.2
+    pass
+
 from collections import defaultdict
 
 

--- a/keys.py
+++ b/keys.py
@@ -1,4 +1,4 @@
-from typing import Dict  # noqa
+from typing import Dict  # noqa: F401
 from collections import defaultdict
 
 

--- a/keys.py
+++ b/keys.py
@@ -1,4 +1,4 @@
-from typing import Dict
+from typing import Dict  # noqa
 from collections import defaultdict
 
 
@@ -8,5 +8,5 @@ def getKey(key):
 
 keys_raw = {
     # add keys here
-} # type: Dict
+}  # type: Dict
 keys = defaultdict(lambda: 'null', keys_raw)

--- a/keys.py
+++ b/keys.py
@@ -1,3 +1,4 @@
+from typing import Dict
 from collections import defaultdict
 
 
@@ -5,6 +6,7 @@ def getKey(key):
     return keys[key]
 
 
-keys = defaultdict(lambda: 'null', {
+keys_raw = {
     # add keys here
-})
+} # type: Dict
+keys = defaultdict(lambda: 'null', keys_raw)

--- a/missingdb.py
+++ b/missingdb.py
@@ -3,6 +3,7 @@
 
 import sqlite3
 import logging
+from typing import Dict
 from datetime import datetime
 import threading
 from collections import defaultdict
@@ -15,7 +16,7 @@ class MissingDb(object):
         self.lock = threading.RLock()
         self.conn = None
         self.dbPath = dbPath
-        self.words = defaultdict(lambda: defaultdict(lambda: 0))
+        self.words = defaultdict(lambda: defaultdict(lambda: 0)) # type: Dict[str, Dict[str, int]]
         self.wordcount = 0
         self.wordmemlimit = wordmemlimit
 

--- a/missingdb.py
+++ b/missingdb.py
@@ -1,9 +1,13 @@
 #!/usr/bin/env python3
 # vim: set ts=4 sw=4 sts=4 et :
 
+try:
+    from typing import Dict  # noqa: F401
+except ImportError:  # 3.2
+    pass
+
 import sqlite3
 import logging
-from typing import Dict  # noqa: F401
 from datetime import datetime
 import threading
 from collections import defaultdict

--- a/missingdb.py
+++ b/missingdb.py
@@ -3,7 +3,7 @@
 
 import sqlite3
 import logging
-from typing import Dict  # noqa
+from typing import Dict  # noqa: F401
 from datetime import datetime
 import threading
 from collections import defaultdict

--- a/missingdb.py
+++ b/missingdb.py
@@ -3,7 +3,7 @@
 
 import sqlite3
 import logging
-from typing import Dict
+from typing import Dict  # noqa
 from datetime import datetime
 import threading
 from collections import defaultdict
@@ -16,7 +16,7 @@ class MissingDb(object):
         self.lock = threading.RLock()
         self.conn = None
         self.dbPath = dbPath
-        self.words = defaultdict(lambda: defaultdict(lambda: 0)) # type: Dict[str, Dict[str, int]]
+        self.words = defaultdict(lambda: defaultdict(lambda: 0))  # type: Dict[str, Dict[str, int]]
         self.wordcount = 0
         self.wordmemlimit = wordmemlimit
 

--- a/modeSearch.py
+++ b/modeSearch.py
@@ -1,7 +1,7 @@
 import re
 import os
 import logging
-from typing import Dict, List, Tuple  # noqa
+from typing import Dict, List, Tuple  # noqa: F401
 from util import toAlpha3Code
 
 

--- a/modeSearch.py
+++ b/modeSearch.py
@@ -1,7 +1,13 @@
+#!/usr/bin/env python3
+
+try:
+    from typing import Dict, List, Tuple  # noqa: F401
+except ImportError:  # 3.2
+    pass
+
 import re
 import os
 import logging
-from typing import Dict, List, Tuple  # noqa: F401
 from util import toAlpha3Code
 
 

--- a/modeSearch.py
+++ b/modeSearch.py
@@ -1,6 +1,7 @@
 import re
 import os
 import logging
+from typing import Dict, List, Tuple
 from util import toAlpha3Code
 
 
@@ -37,7 +38,7 @@ def searchPath(rootpath, include_pairs=True, verbosity=1):
         'analyzer': [],
         'generator': [],
         'tagger': [],
-    }
+    } # type: Dict[str, List[Tuple[str, str, str]]]
 
     real_root = os.path.abspath(os.path.realpath(rootpath))
 

--- a/modeSearch.py
+++ b/modeSearch.py
@@ -1,7 +1,7 @@
 import re
 import os
 import logging
-from typing import Dict, List, Tuple
+from typing import Dict, List, Tuple  # noqa
 from util import toAlpha3Code
 
 
@@ -38,7 +38,7 @@ def searchPath(rootpath, include_pairs=True, verbosity=1):
         'analyzer': [],
         'generator': [],
         'tagger': [],
-    } # type: Dict[str, List[Tuple[str, str, str]]]
+    }  # type: Dict[str, List[Tuple[str, str, str]]]
 
     real_root = os.path.abspath(os.path.realpath(rootpath))
 

--- a/servlet.py
+++ b/servlet.py
@@ -2,6 +2,11 @@
 # -*- indent-tabs-mode: nil -*-
 # coding=utf-8
 
+try:
+    from typing import Dict, List, Optional, Tuple  # noqa: F401
+except ImportError:  # 3.2
+    pass
+
 import sys
 import os
 import re
@@ -14,7 +19,6 @@ import tempfile
 import zipfile
 import string
 import random
-from typing import Dict, List, Optional, Tuple  # noqa: F401
 from subprocess import Popen, PIPE
 from multiprocessing import Pool
 from multiprocessing import TimeoutError  # type: ignore
@@ -50,10 +54,8 @@ import missingdb
 
 if sys.version_info.minor < 3:
     import translation_py32 as translation
-    from translation_py32 import CatPipeline as translation_CatPipeline  # type: ignore
 else:
     import translation  # type: ignore
-    from translation import CatPipeline as translation_CatPipeline  # type: ignore
 
 try:
     import cld2full as cld2  # type: ignore
@@ -741,6 +743,7 @@ class TranslatePageHandler(TranslateHandler):
                 print(e)
                 return
         if got304 and cached is not None:
+            translation_CatPipeline = translation.CatPipeline  # type: ignore
             translated = yield translation_CatPipeline().translate(cached[1])
         else:
             if response.body is None:

--- a/servlet.py
+++ b/servlet.py
@@ -7,14 +7,17 @@ import os
 import re
 import argparse
 import logging
+from logging import handlers as logging_handlers # type: ignore
 import time
 import signal
 import tempfile
 import zipfile
 import string
 import random
+from typing import Dict, List, Optional, Tuple
 from subprocess import Popen, PIPE
-from multiprocessing import Pool, TimeoutError
+from multiprocessing import Pool
+from multiprocessing import TimeoutError # type: ignore
 from functools import wraps
 from threading import Thread
 from datetime import datetime, timedelta
@@ -36,7 +39,7 @@ from tornado.escape import utf8
 try:  # 3.1
     from tornado.log import enable_pretty_logging
 except ImportError:  # 2.1
-    from tornado.options import enable_pretty_logging
+    from tornado.options import enable_pretty_logging # type: ignore
 
 from modeSearch import searchPath
 from keys import getKey
@@ -47,11 +50,13 @@ import missingdb
 
 if sys.version_info.minor < 3:
     import translation_py32 as translation
+    from translation_py32 import CatPipeline as translation_CatPipeline # type: ignore
 else:
-    import translation
+    import translation # type: ignore
+    from translation import CatPipeline as translation_CatPipeline # type: ignore
 
 try:
-    import cld2full as cld2
+    import cld2full as cld2 # type: ignore
 except ImportError as _e:
     cld2 = None
 
@@ -95,12 +100,13 @@ def sig_handler(sig, frame):
 
 
 class BaseHandler(tornado.web.RequestHandler):
-    pairs = {}
-    analyzers = {}
-    generators = {}
-    taggers = {}
-    pipelines = {}  # (l1, l2): [translation.Pipeline], only contains flushing pairs!
-    pipelines_holding = []
+    pairs = {} # type: Dict[str, str]
+    analyzers = {} # type: Dict[str, Tuple[str, str]]
+    generators = {} # type: Dict[str, Tuple[str, str]]
+    taggers = {} # type: Dict[str, Tuple[str, str]]
+    # (l1, l2): [translation.Pipeline], only contains flushing pairs!
+    pipelines = {} # type: Dict[str, List]
+    pipelines_holding = [] # type: List
     callback = None
     timeout = None
     scaleMtLogs = False
@@ -108,11 +114,11 @@ class BaseHandler(tornado.web.RequestHandler):
 
     # dict representing a graph of translation pairs; keys are source languages
     # e.g. pairs_graph['eng'] = ['fra', 'spa']
-    pairs_graph = {}
+    pairs_graph = {} # type: Dict[str, List[str]]
     # 2-D dict storing the shortest path for a chained translation pair
     # keys are source and target languages
     # e.g. paths['eng']['fra'] = ['eng', 'spa', 'fra']
-    paths = {}
+    paths = {} # type: Dict[str, Dict[str, List[str]]]
 
     stats = {
         'startdate': datetime.now(),
@@ -121,7 +127,8 @@ class BaseHandler(tornado.web.RequestHandler):
         'timing': []
     }
 
-    pipeline_cmds = {}  # (l1, l2): translation.ParsedModes
+    # (l1, l2): translation.ParsedModes
+    pipeline_cmds = {} # type: Dict
     max_pipes_per_pair = 1
     min_pipes_per_pair = 0
     max_users_per_pipe = 5
@@ -269,14 +276,12 @@ class ListHandler(BaseHandler):
         if query == 'pairs':
             src = self.get_argument('src', default=None)
             responseData = []
-            if not src:
-                pairs_list = self.pairs
-
-                def langs(pair): return pair.split('-')
-            else:
+            if src:
                 pairs_list = self.paths[src]
-
-                def langs(trgt): return src, trgt
+                langs = lambda lang: (src, lang)
+            else:
+                pairs_list = self.pairs
+                langs = lambda pair: (pair.split('-')[0], pair.split('-')[1])
             for pair in pairs_list:
                 l1, l2 = langs(pair)
                 responseData.append({'sourceLanguage': l1, 'targetLanguage': l2})
@@ -732,7 +737,7 @@ class TranslatePageHandler(TranslateHandler):
                 print(e)
                 return
         if got304 and cached is not None:
-            translated = yield translation.CatPipeline().translate(cached[1])
+            translated = yield translation_CatPipeline().translate(cached[1])
         else:
             if response.body is None:
                 self.send_error(503, explanation="got an empty file on fetching url: {}".format(url))
@@ -1112,7 +1117,7 @@ class GetLocaleHandler(BaseHandler):
 
 
 class SuggestionHandler(BaseHandler):
-    wiki_session = None
+    wiki_session = None # type: Optional[requests.Session]
     wiki_edit_token = None
     SUGGEST_URL = None
     recaptcha_secret = None
@@ -1349,8 +1354,9 @@ if __name__ == '__main__':
         logger = logging.getLogger('scale-mt')
         logger.propagate = False
         smtlog = os.path.join(args.log_path, 'ScaleMTRequests.log')
-        loggingHandler = logging.handlers.TimedRotatingFileHandler(smtlog, 'midnight', 0)
-        loggingHandler.suffix = "%Y-%m-%d"
+        loggingHandler = logging_handlers.TimedRotatingFileHandler(smtlog, 'midnight', 0)
+        # internal attribute, should not use
+        loggingHandler.suffix = "%Y-%m-%d" # type: ignore
         logger.addHandler(loggingHandler)
 
         # if scalemt_logs is enabled, disable tornado.access logs

--- a/servlet.py
+++ b/servlet.py
@@ -14,7 +14,7 @@ import tempfile
 import zipfile
 import string
 import random
-from typing import Dict, List, Optional, Tuple  # noqa
+from typing import Dict, List, Optional, Tuple  # noqa: F401
 from subprocess import Popen, PIPE
 from multiprocessing import Pool
 from multiprocessing import TimeoutError  # type: ignore

--- a/servlet.py
+++ b/servlet.py
@@ -7,17 +7,17 @@ import os
 import re
 import argparse
 import logging
-from logging import handlers as logging_handlers # type: ignore
+from logging import handlers as logging_handlers  # type: ignore
 import time
 import signal
 import tempfile
 import zipfile
 import string
 import random
-from typing import Dict, List, Optional, Tuple
+from typing import Dict, List, Optional, Tuple  # noqa
 from subprocess import Popen, PIPE
 from multiprocessing import Pool
-from multiprocessing import TimeoutError # type: ignore
+from multiprocessing import TimeoutError  # type: ignore
 from functools import wraps
 from threading import Thread
 from datetime import datetime, timedelta
@@ -39,7 +39,7 @@ from tornado.escape import utf8
 try:  # 3.1
     from tornado.log import enable_pretty_logging
 except ImportError:  # 2.1
-    from tornado.options import enable_pretty_logging # type: ignore
+    from tornado.options import enable_pretty_logging  # type: ignore
 
 from modeSearch import searchPath
 from keys import getKey
@@ -50,13 +50,13 @@ import missingdb
 
 if sys.version_info.minor < 3:
     import translation_py32 as translation
-    from translation_py32 import CatPipeline as translation_CatPipeline # type: ignore
+    from translation_py32 import CatPipeline as translation_CatPipeline  # type: ignore
 else:
-    import translation # type: ignore
-    from translation import CatPipeline as translation_CatPipeline # type: ignore
+    import translation  # type: ignore
+    from translation import CatPipeline as translation_CatPipeline  # type: ignore
 
 try:
-    import cld2full as cld2 # type: ignore
+    import cld2full as cld2  # type: ignore
 except ImportError as _e:
     cld2 = None
 
@@ -100,13 +100,13 @@ def sig_handler(sig, frame):
 
 
 class BaseHandler(tornado.web.RequestHandler):
-    pairs = {} # type: Dict[str, str]
-    analyzers = {} # type: Dict[str, Tuple[str, str]]
-    generators = {} # type: Dict[str, Tuple[str, str]]
-    taggers = {} # type: Dict[str, Tuple[str, str]]
+    pairs = {}  # type: Dict[str, str]
+    analyzers = {}  # type: Dict[str, Tuple[str, str]]
+    generators = {}  # type: Dict[str, Tuple[str, str]]
+    taggers = {}  # type: Dict[str, Tuple[str, str]]
     # (l1, l2): [translation.Pipeline], only contains flushing pairs!
-    pipelines = {} # type: Dict[str, List]
-    pipelines_holding = [] # type: List
+    pipelines = {}  # type: Dict[str, List]
+    pipelines_holding = []  # type: List
     callback = None
     timeout = None
     scaleMtLogs = False
@@ -114,11 +114,11 @@ class BaseHandler(tornado.web.RequestHandler):
 
     # dict representing a graph of translation pairs; keys are source languages
     # e.g. pairs_graph['eng'] = ['fra', 'spa']
-    pairs_graph = {} # type: Dict[str, List[str]]
+    pairs_graph = {}  # type: Dict[str, List[str]]
     # 2-D dict storing the shortest path for a chained translation pair
     # keys are source and target languages
     # e.g. paths['eng']['fra'] = ['eng', 'spa', 'fra']
-    paths = {} # type: Dict[str, Dict[str, List[str]]]
+    paths = {}  # type: Dict[str, Dict[str, List[str]]]
 
     stats = {
         'startdate': datetime.now(),
@@ -128,7 +128,7 @@ class BaseHandler(tornado.web.RequestHandler):
     }
 
     # (l1, l2): translation.ParsedModes
-    pipeline_cmds = {} # type: Dict
+    pipeline_cmds = {}  # type: Dict
     max_pipes_per_pair = 1
     min_pipes_per_pair = 0
     max_users_per_pipe = 5
@@ -278,10 +278,14 @@ class ListHandler(BaseHandler):
             responseData = []
             if src:
                 pairs_list = self.paths[src]
-                langs = lambda lang: (src, lang)
+
+                def langs(foo):
+                    return (src, foo)
             else:
                 pairs_list = self.pairs
-                langs = lambda pair: (pair.split('-')[0], pair.split('-')[1])
+
+                def langs(foo):
+                    return (foo.split('-')[0], foo.split('-')[1])
             for pair in pairs_list:
                 l1, l2 = langs(pair)
                 responseData.append({'sourceLanguage': l1, 'targetLanguage': l2})
@@ -1117,7 +1121,7 @@ class GetLocaleHandler(BaseHandler):
 
 
 class SuggestionHandler(BaseHandler):
-    wiki_session = None # type: Optional[requests.Session]
+    wiki_session = None  # type: Optional[requests.Session]
     wiki_edit_token = None
     SUGGEST_URL = None
     recaptcha_secret = None
@@ -1356,7 +1360,7 @@ if __name__ == '__main__':
         smtlog = os.path.join(args.log_path, 'ScaleMTRequests.log')
         loggingHandler = logging_handlers.TimedRotatingFileHandler(smtlog, 'midnight', 0)
         # internal attribute, should not use
-        loggingHandler.suffix = "%Y-%m-%d" # type: ignore
+        loggingHandler.suffix = "%Y-%m-%d"  # type: ignore
         logger.addHandler(loggingHandler)
 
         # if scalemt_logs is enabled, disable tornado.access logs

--- a/tools/sanity-test-apy.py
+++ b/tools/sanity-test-apy.py
@@ -1,13 +1,14 @@
 #!/usr/bin/python3
 
-import urllib.request
 import urllib.parse
+import urllib.request
+import urllib.error
 import socket
 import json
 import sys
 
 import html.parser
-unescape = html.parser.HTMLParser().unescape
+unescape = html.parser.HTMLParser().unescape # type: ignore
 
 TIMEOUT = 15
 tests = {

--- a/tools/sanity-test-apy.py
+++ b/tools/sanity-test-apy.py
@@ -8,7 +8,7 @@ import json
 import sys
 
 import html.parser
-unescape = html.parser.HTMLParser().unescape # type: ignore
+unescape = html.parser.HTMLParser().unescape  # type: ignore
 
 TIMEOUT = 15
 tests = {

--- a/toro.py
+++ b/toro.py
@@ -18,6 +18,7 @@
 import contextlib
 import heapq
 import collections
+from typing import List
 from functools import partial
 from queue import Full, Empty
 
@@ -174,7 +175,7 @@ class AsyncResult(object):
     def __init__(self, io_loop=None):
         self.io_loop = io_loop or ioloop.IOLoop.current()
         self.value = _null_result
-        self.waiters = []
+        self.waiters = [] # type: List[_TimeoutFuture]
 
     def __str__(self):
         result = '<%s ' % (self.__class__.__name__, )
@@ -242,7 +243,8 @@ class Condition(object):
 
     def __init__(self, io_loop=None):
         self.io_loop = io_loop or ioloop.IOLoop.current()
-        self.waiters = collections.deque()  # Queue of _Waiter objects
+        # Queue of _Waiter objects
+        self.waiters = collections.deque() # type: collections.deque
 
     def __str__(self):
         result = '<%s' % (self.__class__.__name__, )
@@ -379,13 +381,13 @@ class Queue(object):
         self._maxsize = maxsize
 
         # _TimeoutFutures
-        self.getters = collections.deque([])
+        self.getters = collections.deque([]) # type: collections.deque
         # Pairs of (item, _TimeoutFuture)
-        self.putters = collections.deque([])
+        self.putters = collections.deque([]) # type: collections.deque
         self._init(maxsize)
 
     def _init(self, maxsize):
-        self.queue = collections.deque()
+        self.queue = collections.deque() # type: collections.deque
 
     def _get(self):
         return self.queue.popleft()

--- a/toro.py
+++ b/toro.py
@@ -15,10 +15,14 @@
 # under the License.
 
 
+try:
+    from typing import List  # noqa: F401
+except ImportError:  # 3.2
+    pass
+
 import contextlib
 import heapq
 import collections
-from typing import List
 from functools import partial
 from queue import Full, Empty
 

--- a/translation.py
+++ b/translation.py
@@ -1,6 +1,12 @@
+#!/usr/bin/env python3
+
+try:
+    from typing import List  # noqa: F401
+except ImportError:  # 3.2
+    pass
+
 import re
 import os
-from typing import List  # noqa: F401
 from subprocess import Popen, PIPE
 from tornado import gen
 import tornado.process

--- a/translation.py
+++ b/translation.py
@@ -1,6 +1,6 @@
 import re
 import os
-from typing import List
+from typing import List  # noqa
 from subprocess import Popen, PIPE
 from tornado import gen
 import tornado.process
@@ -8,7 +8,7 @@ import tornado.iostream
 try:  # >=4.2
     import tornado.locks as locks
 except ImportError:
-    import toro as locks # type: ignore
+    import toro as locks  # type: ignore
 import logging
 from select import PIPE_BUF
 from contextlib import contextmanager
@@ -101,7 +101,7 @@ def makePipeline(modes_parsed):
 
 
 def startPipeline(commands):
-    procs = [] # type: List[tornado.process.Subprocess]
+    procs = []  # type: List[tornado.process.Subprocess]
     for i, cmd in enumerate(commands):
         if i == 0:
             in_from = tornado.process.Subprocess.STREAM

--- a/translation.py
+++ b/translation.py
@@ -1,5 +1,6 @@
 import re
 import os
+from typing import List
 from subprocess import Popen, PIPE
 from tornado import gen
 import tornado.process
@@ -7,7 +8,7 @@ import tornado.iostream
 try:  # >=4.2
     import tornado.locks as locks
 except ImportError:
-    import toro as locks
+    import toro as locks # type: ignore
 import logging
 from select import PIPE_BUF
 from contextlib import contextmanager
@@ -100,7 +101,7 @@ def makePipeline(modes_parsed):
 
 
 def startPipeline(commands):
-    procs = []
+    procs = [] # type: List[tornado.process.Subprocess]
     for i, cmd in enumerate(commands):
         if i == 0:
             in_from = tornado.process.Subprocess.STREAM

--- a/translation.py
+++ b/translation.py
@@ -1,6 +1,6 @@
 import re
 import os
-from typing import List  # noqa
+from typing import List  # noqa: F401
 from subprocess import Popen, PIPE
 from tornado import gen
 import tornado.process

--- a/translation_py32.py
+++ b/translation_py32.py
@@ -1,6 +1,12 @@
+#!/usr/bin/env python3
+
+try:
+    from typing import List  # noqa: F401
+except ImportError:  # 3.2
+    pass
+
 import re
 import os
-from typing import List  # noqa: F401
 from subprocess import Popen, PIPE
 from tornado import gen
 import tornado.process

--- a/translation_py32.py
+++ b/translation_py32.py
@@ -1,6 +1,6 @@
 import re
 import os
-from typing import List
+from typing import List  # noqa
 from subprocess import Popen, PIPE
 from tornado import gen
 import tornado.process
@@ -8,7 +8,7 @@ import tornado.iostream
 try:  # >=4.2
     import tornado.locks as locks
 except ImportError:
-    import toro as locks # type: ignore
+    import toro as locks  # type: ignore
 import logging
 from select import PIPE_BUF
 from contextlib import contextmanager
@@ -101,7 +101,7 @@ def makePipeline(modes_parsed):
 
 
 def startPipeline(commands):
-    procs = [] # type: List[tornado.process.Subprocess]
+    procs = []  # type: List[tornado.process.Subprocess]
     for i, cmd in enumerate(commands):
         if i == 0:
             in_from = tornado.process.Subprocess.STREAM

--- a/translation_py32.py
+++ b/translation_py32.py
@@ -1,5 +1,6 @@
 import re
 import os
+from typing import List
 from subprocess import Popen, PIPE
 from tornado import gen
 import tornado.process
@@ -7,7 +8,7 @@ import tornado.iostream
 try:  # >=4.2
     import tornado.locks as locks
 except ImportError:
-    import toro as locks
+    import toro as locks # type: ignore
 import logging
 from select import PIPE_BUF
 from contextlib import contextmanager
@@ -100,7 +101,7 @@ def makePipeline(modes_parsed):
 
 
 def startPipeline(commands):
-    procs = []
+    procs = [] # type: List[tornado.process.Subprocess]
     for i, cmd in enumerate(commands):
         if i == 0:
             in_from = tornado.process.Subprocess.STREAM

--- a/translation_py32.py
+++ b/translation_py32.py
@@ -1,6 +1,6 @@
 import re
 import os
-from typing import List  # noqa
+from typing import List  # noqa: F401
 from subprocess import Popen, PIPE
 from tornado import gen
 import tornado.process

